Batched (Fork/join parallel) and complex workflows

hillm3 opened the pull request:
Allows for each state of the workflow to query multiple different microservices per thread and also allows for arbitrarily complex workflows.

KamalGalrani commented on lucida/commandcenter/controllers/Config.py:
```diff
@@ -12,41 +12,163 @@
+		if(self.currentState==0):
+			print "State 0";
+			self.currentState = 1;
+			self.batchedData = [serviceRequestData("QA",[unicode("How old is Johann")]),serviceRequestData("QA",[unicode("What color is a pug?")])];
```
Why are these questions hard coded?

hillm3 replied:
That is the demo workflow. That workflow doesn't take any input (Well it takes input, just doesn't use it) and is just to demonstrate how batching and a complex workflow works.

KamalGalrani replied:
Ok

jhauswald commented on lucida/commandcenter/controllers/Config.py:
```diff
@@ -12,41 +12,163 @@
+			self.currentState = 0; # What state on the state graph
+			self.isEnd = False;
+			self.batchedData = []
+	
```
just curious why there's so much space here? are you following a specific style? We should start linting.

hillm3 replied:
I think I should talk to you in person about this.

jhauswald commented on lucida/commandcenter/controllers/Config.py:
```diff
@@ -12,41 +12,163 @@
 # models are already saved in `../models`.
 TRAIN_OR_LOAD = 'train' # either 'train' or 'load'
 
+
```
if this is a demo workflow, should be named as such and maybe in its own directory?

hillm3 replied:
Hm...I think this is something I need to think about in the general proposal I have for Lucida.

Lucida should never even have trained models (Besides the classifier/maybe ASR?). This would be exclusive to the blackboxes.

jhauswald commented on lucida/commandcenter/controllers/Config.py:
```diff
@@ -12,41 +12,163 @@
+
+
+
+class QAWF(workFlow):
```
from my previous comment, is it smart to have all these stateflows here? these are demos but I don't see this file scaling well as more (complex) stateflows get added.

hillm3 replied:
The idea is to have it dynamically stored in a database generated through the GUI tool, so we won't have to look at text at all. Right now it is static, but that should only be temporary. In the future, this will definitely change.

jhauswald commented on lucida/commandcenter/controllers/ThriftClient.py:
```diff
@@ -76,68 +91,71 @@ def learn_text(self, LUCID, text_type, text_data, text_id):
-        self.clear_context(LUCID, service_graph)
-        result = self.send_query(LUCID, service_graph, start_index, query_input_list)
-        return result
+        self.threadResults = []
```
@saijel check the diff here, let me know your thoughts

jhauswald commented:
@zhexuanc can you test this and report your experience?

zhexuanc commented:
I have tested it and there are some issues.

- QA seems broken. I look through the source code and error and it is because the there are still dependency in two microservices so the query is not parsed correctly.
- IMC, DIG, FACE always show "Not a JPEG file: starts with 0x5b 0x5b"
- IMM is not connected. The backend service does not have any response for query. 

I am not sure if this is only a test version, or we need to fixes service based on the commandcenter.

hillm3 commented:
QA had an issue from the previous workflow system that is now deleted.
IMC/DIG/FACE were being passed as text in the workflow, not image. That is changed.
IMM somehow I forgot to give a WF to.

Let me know if any of the issues still persist.

zhexuanc commented:
I test it and the QA/IMC/DIG/FACE is now working. 
For IMM the backend service breaks down when querying. Also, I am not clear how to add my own workflow and how to load the classification data of my workflow into lucida.

hillm3 commented:
I'll add some documentation. Classifier should be the same thing as I hadn't changed that.

Looking at IMM, I think the same issue with QA again. IMM is suffering from the old workflow system.

hillm3 commented:
IMM fixed.

Added some documentation. In the future, the GUI will be used to avoid the need to read documentation. Also removed some bad commented code.
